File: ruby/005.rb
# Smallest multiple
# Problem 5 
# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
# 
# What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

# for all multiples of 20, check if the other numbers are also a mutliple

require 'prime'

primeDivisions = (2..20).map do |n|
  Prime::prime_division(n)
end

lcmMap = {}

primeDivisions.each do |primeCount|
  primeCount.each do |prime, count|
    if (lcmMap[prime] != nil && lcmMap[prime] < count) || (lcmMap[prime] == nil)
      lcmMap[prime] = count
    end
  end
end
lcm = lcmMap.inject(1) do |acc, primeCount|
  prime = primeCount[0]
  count = primeCount[1]
  acc * prime ** count
end
puts lcm

File: haskell/005.hs
-- Smallest multiple
-- Problem 5 
-- 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
-- 
-- What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
--
-- LCM of [1..20]

import qualified Data.Map.Strict as Map
import Data.Numbers.Primes
import Data.List

baseMap :: Map.Map Int Int
baseMap = Map.empty

numbers = [2..20]

main = do
  let primeFactorLists = map primeFactors numbers
  let primeFactorMaps = map generatePrimeFactorMap primeFactorLists
  let lcmMap = Map.unionsWith greater primeFactorMaps
  print $ Map.foldrWithKey (\k val acc -> acc * (k ^ val)) 1 lcmMap

checkMapAndReturnFinalValue :: Map.Map Int Int -> Int -> Int -> Map.Map Int Int
checkMapAndReturnFinalValue primePowerMap prime power = Map.insert prime finalValue primePowerMap
  where maybePower = Map.lookup prime primePowerMap
        finalValue = check maybePower power
        check (Just a) b = if a > b then a else b
        check Nothing b = b

generatePrimeFactorMap primeFactorsList = primeFactorMap
  where grouped = group primeFactorsList
        primeFactorMap = foldr (\val acc -> Map.insert (head val) (length val) acc) Map.empty grouped

greater = (\a b -> if a > b then a else b)
